--- /Users/gabrielhicks/Staking/jito-mod/sdk/program/src/vote/state/mod.rs	2024-10-07 14:40:52
+++ /Users/gabrielhicks/Staking/jito-mod/jito-zan-ax/mod_files/sdk/program/src/vote/state/mod.rs	2024-10-07 14:43:03
@@ -453,6 +453,7 @@
         current_slot: Slot,
         timely_vote_credits: bool,
         deprecate_unused_legacy_vote_plumbing: bool,
+        pop_expired: bool,
     ) {
         // Ignore votes for slots earlier than we already have votes for
         if self
@@ -462,7 +463,9 @@
             return;
         }
 
-        self.pop_expired_votes(next_vote_slot);
+        if pop_expired {
+            self.pop_expired_votes(next_vote_slot);
+        }
 
         let landed_vote = LandedVote {
             latency: if timely_vote_credits || !deprecate_unused_legacy_vote_plumbing {
