--- /Users/gabrielhicks/Staking/jito-mod/core/src/consensus/progress_map.rs	2024-10-07 14:40:52
+++ /Users/gabrielhicks/Staking/jito-mod/jito-zan-ax/mod_files/core/src/consensus/progress_map.rs	2024-10-07 14:43:03
@@ -303,6 +303,7 @@
     pub is_locked_out: bool,
     pub voted_stakes: VotedStakes,
     pub is_supermajority_confirmed: bool,
+    pub is_mostly_confirmed: bool,
     pub computed: bool,
     pub lockout_intervals: LockoutIntervals,
     pub bank_hash: Option<Hash>,
@@ -487,6 +488,11 @@
     pub fn set_supermajority_confirmed_slot(&mut self, slot: Slot) {
         let slot_progress = self.get_mut(&slot).unwrap();
         slot_progress.fork_stats.is_supermajority_confirmed = true;
+    }
+
+    pub fn set_mostly_confirmed_slot(&mut self, slot: Slot) {
+        let slot_progress = self.get_mut(&slot).unwrap();
+        slot_progress.fork_stats.is_mostly_confirmed = true;
     }
 
     pub fn is_supermajority_confirmed(&self, slot: Slot) -> Option<bool> {
