--- /Users/gabrielhicks/Staking/jito-mod/core/src/vote_simulator.rs	2024-10-07 14:40:52
+++ /Users/gabrielhicks/Staking/jito-mod/jito-zan-ax/mod_files/core/src/vote_simulator.rs	2024-10-07 14:43:03
@@ -183,6 +183,7 @@
 
         // Try to vote on the given slot
         let descendants = self.bank_forks.read().unwrap().descendants();
+        let mut last_logged_vote_slot = 0;
         let SelectVoteAndResetForkResult {
             heaviest_fork_failures,
             ..
@@ -195,6 +196,7 @@
             tower,
             &self.latest_validator_votes_for_frozen_banks,
             &self.heaviest_subtree_fork_choice,
+            &mut last_logged_vote_slot,
         );
 
         // Make sure this slot isn't locked out or failing threshold
@@ -203,7 +205,7 @@
             return heaviest_fork_failures;
         }
 
-        let new_root = tower.record_bank_vote(&vote_bank);
+        let new_root = tower.record_bank_vote(&vote_bank, true);
         if let Some(new_root) = new_root {
             self.set_root(new_root);
         }
